### PR TITLE
8335939: Hide element writing across the ClassFile API

### PR DESCRIPTION
**Pull request by liach**
`WritableElement` has always been one of the biggest peculiarities of ClassFile API: it exposes element writing yet has no corresponding reading ability exposed. Its existence creates a lot of API noise, increasing maintenance cost in the long run. (This is an iceberg whose tip was exposed in #14705)

Removal details:
- `LocalVariable/LocalVariableType.writeTo`
- `WritableElement`
- In `BufWriter`:
  - `writeList(List<WritableElement>)`
  - `writeListIndices`: Hidden as we are not exposing `BoundAttribute.readEntryList`
  - `writeBytes(BufWriter other)`
- `ClassReader.compare`: Avoid reading from `BufWriter`

Future implementation cleanup out of scope of this patch:
- Annotation writing can be upgraded and move away from `Util.Writable`
- The writing of CP indices and attributes can move to their dedicated methods
- reading of entry list can pair up with writing of entry list in ClassReader/BufWriter in future addition

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8336586](https://bugs.openjdk.org/browse/JDK-8336586) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8335939](https://bugs.openjdk.org/browse/JDK-8335939): Hide element writing across the ClassFile API (**Enhancement** - P3)
 * [JDK-8336586](https://bugs.openjdk.org/browse/JDK-8336586): Remove WritableElement and hide writing details of BufWriter (**CSR**)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20205/head:pull/20205` \
`$ git checkout pull/20205`

Update a local copy of the PR: \
`$ git checkout pull/20205` \
`$ git pull https://git.openjdk.org/jdk.git pull/20205/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20205`

View PR using the GUI difftool: \
`$ git pr show -t 20205`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20205.diff">https://git.openjdk.org/jdk/pull/20205.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20205#issuecomment-2232028714)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java [hunk 1]
```diff
@@ -280,4 +281,24 @@ private static int skipTypeAnnotation(ClassReader classReader, int p) {
         p = skipElementValuePairs(classReader, p);
         return p;
     }
+
+    public static void writeAnnotation(BufWriterImpl buf, Annotation annotation) {
+        // handles annotations and type annotations
+        // TODO annotation cleanup later
```
Comment (asotona) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java:287]: Do you have any specific annotation cleanup in mind?

**Review by asotona — APPROVED** [review]
Great job!

File: src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java:287]: I am thinking of removing the Writable Hierarchy from annotations, and these static methods will handle writing with like a switch (on char, can't use pattern match in bootstrap) instead

**Review by asotona — APPROVED** [review]

**Comment by liach** [comment]
Re-requesting a review: there was a conflict with #20244 in implements list of `DirectMethodBuilder`.

**Review by asotona — APPROVED** [review]

**Comment by liach** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**